src/Symfony/Bundle/FrameworkBundle/Client.php

```diff
@@ -104,6 +106,34 @@ public function enableReboot()
         $this->reboot = true;
     }
 
+    /**
+     * Logins a user against a specific firewall.
+     *
+     * @param string   $username
+     * @param string   $password
+     * @param string   $firewall
+     * @param string[] $roles
+     */
+    public function login($username, $password, $firewall = 'main', $roles = array('ROLE_ADMIN'))
+    {
+        $session = $this->getContainer()->get('session');
+
+        $token = new UsernamePasswordToken($username, $password, $firewall, $roles);
+        $session->set('_security_'.$firewall, serialize($token));
```

Comment:
We have a little problem with the `$firewall` argument. We recently updated the related page (http://symfony.com/doc/current/testing/http_authentication.html) to show that this code sometimes needs the `$firewallName` and sometimes, the `$firewallContext`. Both can or cannot be the same depending on each app.

I can think of 2 alternatives:

1) We could rename `$firewall` to `$firewallName` and don't support the case of different firewall names and context.
2) Update this method signature as follows:

```php
// Before
login(
    string $username,
    string $password,
    string $firewall = 'main',
    array $roles = array('ROLE_ADMIN')
);

// After
login(
    string $username,
    string $password,
    array $roles = array('ROLE_ADMIN')
    string $firewallName = 'main',
    string $firewallContext = null,
) {
    // ...
    $context = $firewallContext ?? $firewallName;
}
```

Reply:
Let's go for the 2nd option 👍

Reply:
Fixed

Reply:
Having to remember this many arguments is not good for DX IMO, maybe we can use context builder of sorts?

```php
$client->loginWith(UsernamePassword($username, $password, $roles)->to($firewallName, $firewallContext));
```

While still allowing the simpler version:

```php
$client->login($username, $password, $roles);
```

We could even allow different implementations of `LoginContext` when you need to test authentication with tokens for example. Making this method more generic in purpose.

Btw. If you pass the firewall-name shouldn't this read the configuration of the application? Rather then having to pass the context, which could be different in the actual application config, making your test pass but fail in a real-life situation.